mockrxandroidble/src/main/java/com/polidea/rxandroidble2/mockrxandroidble/RxBleConnectionMock.java

```diff
@@ -144,29 +156,71 @@ RxBleDeviceServices getRxBleDeviceServices() {
 
     @Override
     public Single<byte[]> readCharacteristic(@NonNull UUID characteristicUuid) {
-        return getCharacteristic(characteristicUuid).map(new Function<BluetoothGattCharacteristic, byte[]>() {
+        final Function<BluetoothGattCharacteristic, Single<byte[]>> readCallback = characteristicReadCallbacks.get(characteristicUuid);
+        Single<BluetoothGattCharacteristic> characteristic = getCharacteristic(characteristicUuid);
+        if (readCallback == null) {
+            return characteristic.map(new Function<BluetoothGattCharacteristic, byte[]>() {
+                @Override
+                public byte[] apply(BluetoothGattCharacteristic bluetoothGattCharacteristic) throws Exception {
+                    return bluetoothGattCharacteristic.getValue();
+                }
+            });
+        }
+        return characteristic.flatMap(new Function<BluetoothGattCharacteristic, SingleSource<? extends byte[]>>() {
             @Override
-            public byte[] apply(BluetoothGattCharacteristic bluetoothGattCharacteristic) throws Exception {
-                return bluetoothGattCharacteristic.getValue();
+            public SingleSource<? extends byte[]> apply(final BluetoothGattCharacteristic characteristic) throws Exception {
+                return Single.just(characteristic).flatMap(readCallback).doOnSuccess(new Consumer<byte[]>() {
+                    @Override
+                    public void accept(byte[] bytes) throws Exception {
+                        characteristic.setValue(bytes);
+                    }
+                });
             }
         });
```

Comment:
This looks like it could be:
```
Single<BluetoothGattCharacteristic> characteristic = getCharacteristic(characteristicUuid);
return readCharacteristic(characteristic);
```
Doesn't it?

Similarly in other places